src/main/java/com/google/cloud/tools/managedcloudsdk/internal/download/StandardDownloader.java

```diff
@@ -0,0 +1,110 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NotDirectoryException;
+import java.nio.file.Path;
+
+/** Standard implementation of {@link Downloader}. */
+public class StandardDownloader implements Downloader {
+
+  static final int BUFFER_SIZE = 10 * 1024;
+  private final URL address;
+  private final Path destinationFile;
+  private final String userAgentString;
+  private final DownloadProgressListener downloadProgressListener;
+
+  /** Use {@link DownloaderFactory} to instantiate. */
+  StandardDownloader(
+      URL source,
+      Path destinationFile,
+      String userAgentString,
+      DownloadProgressListener downloadProgressListener) {
+    this.address = source;
+    this.destinationFile = destinationFile;
+    this.userAgentString = userAgentString;
+    this.downloadProgressListener = downloadProgressListener;
+  }
+
+  @Override
+  public Path call() throws IOException, InterruptedException {
+
+    createDestinationDirectory();
+    return downloadUrl();
+  }
+
+  @VisibleForTesting
+  void createDestinationDirectory() throws IOException {
+    Path destinationDir = destinationFile.getParent();
+    if (Files.exists(destinationDir) && !Files.isDirectory(destinationDir)) {
+      throw new NotDirectoryException(
+          "Cannot download to " + destinationDir + " because it is not a directory");
+    }
+    if (Files.exists(destinationFile)) {
+      throw new FileAlreadyExistsException(
+          "Cannot write to " + destinationFile + " because it already exists");
+    }
+    if (!Files.exists(destinationDir)) {
+      Files.createDirectories(destinationDir);
+    }
+  }
+
+  @VisibleForTesting
+  Path downloadUrl() throws IOException, InterruptedException {
+    URLConnection conn;
+    conn = address.openConnection();
+    conn.setRequestProperty("User-Agent", userAgentString);
+
+    try (BufferedOutputStream out =
+        new BufferedOutputStream(Files.newOutputStream(destinationFile))) {
+      try (InputStream in = conn.getInputStream()) {
+
+        long contentLength = conn.getContentLengthLong();
+        long updateThreshold = contentLength / 100;
+
+        byte[] buffer = new byte[BUFFER_SIZE];
+        long lastUpdated = 0;
+        long totalBytesRead = 0;
+
+        int bytesRead;
+        while ((bytesRead = in.read(buffer)) != -1) {
+          if (Thread.currentThread().isInterrupted()) {
+            // throw exception here? todo?
+            throw new InterruptedException("Downloader was interrupted.");
+          }
+          totalBytesRead += bytesRead;
+          if (totalBytesRead - lastUpdated > updateThreshold) {
+            lastUpdated = totalBytesRead;
+            if (downloadProgressListener != null) {
+              downloadProgressListener.updateProgress(bytesRead, totalBytesRead, contentLength);
```

Comment:
According the the annotation for `DownloadProgressListener#updateProgress`, the first parameter is `@param lastChunk bytes read since last update.` Should this be what the value of `totalBytesRead - lastUpdated` was rather than `bytesRead`?

Reply:
good call, I added that in late, and my test didn't really account for it. :\